notifier/eventsHandler.go

```diff
@@ -0,0 +1,145 @@
+package notifier
+
+import (
+	"context"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go-logger/check"
+	"github.com/ElrondNetwork/notifier-go/config"
+	"github.com/ElrondNetwork/notifier-go/data"
+	"github.com/ElrondNetwork/notifier-go/redis"
+)
+
+const (
+	setnxRetryMs       = 500
+	revertKeyPrefix    = "revert_"
+	finalizedKeyPrefix = "finalized_"
+)
+
+// ArgsEventsHandler defines the arguments needed for an events handler
+type ArgsEventsHandler struct {
+	Config    config.ConnectorApiConfig
+	Locker    redis.LockService
+	Publisher Publisher
+}
+
+type eventsHandler struct {
+	config    config.ConnectorApiConfig
+	locker    redis.LockService
+	publisher Publisher
+}
+
+// NewEventsHandler creates a new events handler component
+func NewEventsHandler(args ArgsEventsHandler) (*eventsHandler, error) {
+	err := checkArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	return &eventsHandler{
+		locker:    args.Locker,
+		publisher: args.Publisher,
+	}, nil
+}
+
+func checkArgs(args ArgsEventsHandler) error {
+	if check.IfNil(args.Locker) {
+		return ErrNilLockService
+	}
+	if check.IfNil(args.Publisher) {
+		return ErrNilPublisherService
+	}
+
+	return nil
+}
+
+// HandlePushEvents will handle push events received from observer
+func (eh *eventsHandler) HandlePushEvents(events data.BlockEvents) {
+	shouldProcessEvents := true
+	if eh.config.CheckDuplicates {
+		shouldProcessEvents = eh.tryCheckProcessedOrRetry(events.Hash)
+	}
+
+	if events.Events != nil && shouldProcessEvents {
+		log.Info("received events for block",
+			"block hash", events.Hash,
+			"shouldProcess", shouldProcessEvents,
+		)
+		eh.publisher.BroadcastChan() <- events
```

Comment:
if eh.publisher.BroadcastChan() will return a nil chan the code will panic in a deadlock. Also for a closed chan.

Strongly suggest to change all these chan returning functions into something like:
```GO
eh.publisher.Broadcast(events)
```

and internally, in the publisher, decide if you will write on a chan or do nothing (in the case of disabled elements)

Valid for all methods returning a chan on which you want to write something

Reply:
refactored with the suggestion, thanks :+1:

Reply:
since this setup was already implemented based on `dispatcher.Hub` interface, i modified here (for now) in publisher (from rabbitmq) implementation which i've decoupled. For hub components will do this later, that part was not refactored yet, will add a TODO for that, it will be done in the following PRs

Reply:
refactored Hub interface in this PR also